Implement stateDiff tracer, compatible with OpenEthereum

ziogaschr opened the pull request:
This PR adds **stateDiff** tracer that is compatible with OpenEthereum.

> **stateDiff State difference**. Provides information detailing all altered portions of the Ethereum state made due to the execution of the transaction.

The **stateDiff** tracer can be used in the following **Ad-hoc Tracing** methods:

- [x] trace_call
- [x] trace_callMany
- [ ] trace_rawTransaction
- [ ] trace_replayBlockTransactions
- [ ] trace_replayTransaction

p.s.: The remaining unchecked **Ad-hoc Tracing** methods will be added in a new PR, and will have stateDiff tracer support out of the box.

### Differences with OpenEthereum

1. SSTORE in some edge cases are being persisted in state but are being removed from stateDiff storage results. (Happens only on 2 transactions on Mordor testnet)
2. When error ErrInsufficientFundsForTransfer happens, OpenEthereum leaves the tracer run producing negative balances, though using safe math for overflows it returns 0 balance, on the other hand the **to** account receives the full amount. Core-geth removes only the gas cost from the **sender** and adds it to the **coinbase** balance.
3. Same as in 2, but on top of that, the **sender** account doesn't have to pay for the gas cost even. In this case, core-geth returns an empty JSON, as in reality this transaction will remain in the tx_pool and never be executed, neither change the state.
4. On OpenEthereum the **block gasLimit** is set to be [`U256::max()`](https://github.com/openethereum/openethereum/blob/main/crates/ethcore/src/client/client.rs#L1857), which leads into problems on contracts using it for pseudo-randomness. On core-geth, we believe that the user utilising the `trace_*` wants to see what will happen in reality, though we leave the block untouched to its true values.
5. When an internal call fails with out of gas, and its state is not being persisted, we don't add it in stateDiff output, as it happens on OpenEthereum.

ziogaschr committed:
eth: extract traceChain from the PrivateDebugAPI

ziogaschr committed:
eth: add initial draft API for `trace_filter`

ziogaschr committed:
eth: add `trace_call`

ziogaschr committed:
eth: tracer_call to allow handling other tracers

ziogaschr committed:
eth/tracers: add draft state_diff JS tracer

eth: add test for stateDiff tracer

ziogaschr committed:
eth: add trace_callMany

ziogaschr committed:
eth: add block context for trace_call

ziogaschr committed:
core/vm, eth/tracers: add EVM param in CaptureEnd

ziogaschr committed:
🍬 cleanup

ziogaschr committed:
eth/tracers: add JS context for statedb.Empty

ziogaschr committed:
eth/tracers: work on state_diff tracer

eth, eth/tracers: resolve state_diff misconsistencies

eth: pass extra context to call JS tracers hasFromSufficientBalance*

ziogaschr committed:
eth/tracers: add support for init method call on supported JS tracers

ziogaschr committed:
eth/tracers: work on balance calculations

ziogaschr committed:
eth: trace_call to run on top of the provided block (with txs) on reexec

ziogaschr committed:
eth, eth/tracers: cleanup/verify state_diff logic

eth: traceCall to use original TransferFunc instead of copy the logic and take care of mutating the value

ziogaschr committed:
eth, eth/tracers: try complex code cleaning for state_diff by adding CapturePreEVM

Till this point, state_diff was manually calculating balance/nonce changes that have been executed pre/post EVM execution. This was error prone, as any change in consensus within go, will have to be transfered to JS logic as well.

This commit tackles this issue, by adding an init() method call, before anything is be run for a transaction.

eth: clean up vmctx.Tracer (pointer, refs, oh my)

Date: 2021-03-12 15:09:48-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
eth, eth/tracers: fix trace_callMany for state_diff; will deduplicate code after the upstream merge

ziogaschr committed:
eth/tracers: add state_diff JS tracer tests

ziogaschr committed:
eth/tracers: update prestate tracer to use the new CapturePreEVM/init() functionality

ziogaschr committed:
eth/tracers: remove debug logs from state_diff

ziogaschr committed:
eth/tracers: add stateDiff test for internall call OOG

ziogaschr committed:
eth/tracers: refactor the test runner for state_diff

ziogaschr committed:
Merge branch 'master' into feat/trace-state-diff

ziogaschr committed:
eth, eth/tracers: fixes

eth/tracers: remove commented out forgoten code

eth/tracers: remove unused types from tests

eth: fix lint issue :(

eth: remove spurious comment

ziogaschr commented on eth/api_tracer.go:
```diff
@@ -824,16 +914,65 @@ func (api *PrivateDebugAPI) TraceCall(ctx context.Context, args ethapi.CallArgs,
-	msg := args.ToMessage(api.eth.APIBackend.RPCGasCap())
-	vmctx := core.NewEVMContext(msg, header, api.eth.blockchain, nil)
-	return api.traceTx(ctx, msg, vmctx, statedb, config)
+	var results = make([]interface{}, len(txs))
```
The below code is exact copy of the one in `TraceCall` which can be decoupled on its own function. I kept it here until we merge from go-ethereum and we take the refactor where they moved the trace stuff into a package.
I will refactor the logic then on a new PR.

ziogaschr replied:
Also the state_diff tests will need this logic. At the moment, we have a 3rd partial copy there.

ziogaschr commented on eth/tracers/internal/tracers/prestate_tracer.js:
```diff
@@ -18,7 +18,8 @@
 {
 	// prestate is the genesisT that we're building.
-	prestate: null,
+	prestate: {},
```
This tracer, is being used to get the prestate for tests to be used as genesis. I changed this as well to utilise the CapturePreEVM as well.

ziogaschr committed:
eth, eth/tracers, ethclient: cleanup, comments, semi-fix test

ziogaschr committed:
eth/tracers: add stateDiff test for a slow tx

ziogaschr committed:
ethclient: fix discovery test by removing subscription based “trace_filter”

meowsbits commented on ethclient/ethclient_test.go:
```diff
@@ -809,7 +809,8 @@ var allRPCMethods = []string{
 	"personal_unlockAccount",
 	"personal_unpair",
 	"trace_block",
-	"trace_filter",
```
Why is `trace_filter` removed?

ziogaschr replied:
Because it is subscription based.

meowsbits replied:
I'm pretty sure there's a way to keep this in the docs/OpenRPC generation. 

`eth_subscribe` establishes subscriptions with a different signature pattern (eg. `eth_subscribe ["newHeads"]` -> `(*api.Eth) SubscribeNewHeads(...) (rpc.Subscription, error)`.
So if this convoluted pattern can do it, then `trace_filter` must be able to too :wink:. 

It might require some noodling in node/openrpc.go.

ziogaschr replied:
That's great. I will have a look at it. 
Do you think it worths handling this in a next PR, so as we merge this one? As we have to add the remaining bits for `trace_fitler` as well in a next PR.

ziogaschr replied:
@meowsbits can you please let me know if this is what you had in mind? #345

ziogaschr committed:
eth: allow Parity Ad-hoc tracers to be nested for full compatiblity with Parity

meowsbits commented on eth/api_tracer_parity.go:
```diff
@@ -71,6 +71,21 @@ func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
 	return config
 }
 
+func enchanceParityOutput(res interface{}, config *TraceConfig) interface{} {
```
Typo here `s/enchance/enhance`. Maybe `decorateParityOutput` instead?

ziogaschr committed:
eth: fix typo

meowsbits commented on eth/api_tracer_parity.go:
```diff
@@ -71,6 +71,21 @@ func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
 }
 
+func enchanceParityOutput(res interface{}, config *TraceConfig) interface{} {
+	if config.ParityCompatibleOutput {
```
For this logic, I would recommend moving this condition out of the function (eg. to line 252). 

```
if shouldDecorate {
    decorate()
}
```

vs. 

```
decorate() {
    if !shouldDecorate {
        return
    }
}

ziogaschr replied:
Thanks man, handled at 3582231. You triggered me to add support for all available JS tracers as well. 👍

meowsbits replied:
Added another touch to it - https://github.com/etclabscore/core-geth/pull/345/commits/a5d00e081c2427a26778bd7f66789da6bab8ce6a

Per the commit message, up for your review too

ziogaschr replied:
@meowsbits I am fine with keeping the a5d00e08, though I left it as it was as I am afraid of future dev mistakes that can lead to a bug by not changing the if condition in all places. Reason is that this exact call/check will be called from 5 (and maybe even 8) different places/functions once all the trace methods are being supported.

Thinking it twice and based on where you placed the check now, by doing a slight modification on `setConfigTracerToParity` we can change the condition to be `if config.ParityCompatibleOutput {` which makes much more sense then and it's harder to break the logic on future. Will check if it makes sense.

ziogaschr replied:
Resolved with 84bf302

meowsbits commented on eth/api_tracer.go:
```diff
@@ -60,9 +60,10 @@ const (
+	Tracer                 *string
+	Timeout                *string
+	Reexec                 *uint64
+	ParityCompatibleOutput bool
```
Does `TraceConfig` have custom JSON Un/Marshal methods? Where/how are these fields configured? (Forgive my naivete... :blush: )

ziogaschr replied:
No, it doesn't. It just uses the default field names as the name. They just fallback to the default value for the type at the moment, or if no key is set, then the whole config is nil.

I was thinking to make this explicit when I joined ETC and started working on Tracers, but left it like this to not trigger any future conflicts with go-ethereum.

ziogaschr replied:
Related, here I initiate it using the default if needed. https://github.com/etclabscore/core-geth/blob/feat/trace-state-diff/eth/api_tracer_parity.go#L66

meowsbits replied:
Got it. For ease of reference

```go
// setConfigTracerToParity forces the Tracer to the Parity one
func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
	if config == nil {
		config = &TraceConfig{}
	}

	if config.Tracer == nil {
		tracer := "callTracerParity"
		config.Tracer = &tracer
	}

	return config
}
```

Used as 

```go
// Block returns the structured logs created during the execution of
// EVM and returns them as a JSON object.
// The correct name will be TraceBlockByNumber, though we want to be compatible with Parity trace module.
func (api *PrivateTraceAPI) Block(ctx context.Context, number rpc.BlockNumber, config *TraceConfig) ([]interface{}, error) {
	// Fetch the block that we want to trace
	var block *types.Block

	switch number {
	case rpc.PendingBlockNumber:
		block = api.eth.miner.PendingBlock()
	case rpc.LatestBlockNumber:
		block = api.eth.blockchain.CurrentBlock()
	default:
		block = api.eth.blockchain.GetBlockByNumber(uint64(number))
	}
	// Trace the block if it was found
	if block == nil {
		return nil, fmt.Errorf("block #%d not found", number)
	}

	config = setConfigTracerToParity(config)

	traceResults, err := traceBlockByNumber(ctx, api.eth, number, config)
	if err != nil {
		return nil, err
	}
```

In this case it seems to me like the `setConfigTracerToParity` is actually _defaulting_ the config to Parity in case the value is nil (as opposed to `forces the Tracer to the Parity one`), or? Maybe just a rename for that function?

meowsbits replied:
I'm curious about the JSON field encoding here. Normally JSON fields for exported methods will use capitalized field names;

```go
//cat main.go
package main

import (
        "encoding/json"
        "fmt"
)

type MyStruct struct {
        Reexec             *string
        ParityCompatOutput bool
}

func main() {
        ex := &MyStruct{}

        b, _ := json.MarshalIndent(ex, "", "    ")
        fmt.Println(string(b))
}
```

```sh
go run main.go
{
    "Reexec": null,
    "ParityCompatOutput": false
}
```

And in the internal/web3ext.go it doesn't look like this arg is getting a custom formatter. (~~Also, should `callMany` have same formatters as `call`?~~ No, because args are different; callMany uses an array of CallArgs. Should this have its own formatter?).  

```js
		new web3._extend.Method({
			name: 'call',
			call: 'trace_call',
			params: 3,
			inputFormatter: [web3._extend.formatters.inputCallFormatter, null, null]
		}),
		new web3._extend.Method({
			name: 'callMany',
			call: 'trace_callMany',
			params: 3,
			inputFormatter: [null, null, null]
		}),
```

ziogaschr replied:
> In this case it seems to me like the setConfigTracerToParity is actually defaulting the config to Parity in case the value is nil (as opposed to forces the Tracer to the Parity one), or? Maybe just a rename for that function?

Good catch. It's a comment issue. It has been left off, from the initial tracing work. Will be updated.

> I'm curious about the JSON field encoding here. Normally JSON fields for exported methods will use capitalised field names;

True. Currently this is only used to Unmarshal JSON and the default parser handles `camelCase` as well. I followed go-ethereum practice without paying much attention to this one and you are right that this can create issues later on.
Do you think it's good to handle this now? and if yes, shall we do this in this or another PR?

> And in the internal/web3ext.go it doesn't look like this arg is getting a custom formatter. (Also, should callMany have same formatters as call? No, because args are different; callMany uses an array of CallArgs. Should this have its own formatter?).

Yes, this is a great idea. Though we have to check out how the `internal/jsre/deps/web3.js` is being packaged externally, and if such change will be easy to maintain (conflicts resolution) later on.
@meowsbits idea to see if we can extend this functionality in `web3ext.go` is great. I will examine this, as I have some concerns with regards the custom function context.

ziogaschr replied:
Resolved by e0b8ae0 and 84bf302.

meowsbits replied:
Nicely done with https://github.com/etclabscore/core-geth/pull/345/commits/e0b8ae0ce39231bf76229dbbf5142aa496bdd139!

meowsbits replied:
> I followed go-ethereum practice without paying much attention to this one and you are right that this can create issues later on. Do you think it's good to handle this now? and if yes, shall we do this in this or another PR?

Understood. As long as there are tests ensuring that all the fields of the Config get decoded as expected, then I think it's fine as-is.

ziogaschr committed:
docs: add trace module overview

ziogaschr committed:
eth: Allow `trace_*` methods to use all the available JS tracers

meowsbits commented on eth/tracers/tracer.go:
```diff
@@ -561,6 +568,25 @@ func wrapError(context string, err error) error {
 	return fmt.Errorf("%v    in server-side tracer function '%v'", err, context)
 }
 
+func (jst *Tracer) CapturePreEVM(env *vm.EVM, inputs map[string]interface{}) error {
```
Other tracer `Capture` methods are defined as interface methods in core/vm/logger.go's `Tracer` type. This interface is implemented by `JSONLogger`, `StructLogger`, and `Tracer`. Should `CapturePreEVM` be likewise defined in that interface? Why or why not?

ziogaschr replied:
Initially, I left it out of the interface as it will only be used for the JS tracers and not the rest.
On the other hand is a good addition, and later on rest tracers might have some benefits using it.
So bbc4a7b adds `CapturePreEVM` in the interface, to make everything cleaner.

Thanks @meowsbits.

meowsbits commented on docs/JSON-RPC-API/trace-module-overview.md:
```diff
@@ -0,0 +1,53 @@
+
+### Available tracers
+
+- `callTracerParity` Transaction trace.
```
```suggestion
- `callTracerParity` Transaction trace returning a response equivalent to OpenEthereum's (aka Parity) response schema. For documentation on this response value see _______.
```

ziogaschr replied:
Done at 2e486f6

meowsbits commented on docs/JSON-RPC-API/trace-module-overview.md:
```diff
@@ -0,0 +1,53 @@
+# "trace" Module Overview
+
+The trace module is for getting a deeper insight into transaction processing. It includes two sets of calls; the transaction trace filtering API and the ad-hoc tracing API. You can find the documentation for the supported methods [here](/JSON-RPC-API/modules/trace/).
```
Some context here regarding the relationship of methods in `trace_` which may be similar to ones in `debug_` would, I think, be helpful.

ziogaschr replied:
Done at 2e486f6

meowsbits commented on docs/JSON-RPC-API/trace-module-overview.md:
```diff
@@ -0,0 +1,53 @@
+
+## "stateDiff" tracer differences with OpenEthereum
+
+1. **SSTORE** in some edge cases persists data in state but are not being returned on stateDiff storage results on OpenEthereum output. *(Happens only on 2 transactions on Mordor testnet)*.
```
Citations of canonical chain data should probably include state information, like _as of block 1,500,214_.

ziogaschr replied:
Done at 2e486f6

meowsbits commented on core/vm/interpreter.go:
```diff
@@ -254,8 +254,7 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 		if operation.dynamicGas != nil {
-			var dynamicCost uint64
-			dynamicCost, err = operation.dynamicGas(in.evm, contract, stack, mem, memorySize)
+			dynamicCost, err := operation.dynamicGas(in.evm, contract, stack, mem, memorySize)
```
This will assign two new variables instead of only one (reassigning the return value `err`). Is this just a pretty-fication change?

ziogaschr replied:
Reverted 53eaf17. Great catch, as it will trigger issues.

meowsbits commented on eth/api_tracer.go:
```diff
@@ -140,7 +142,7 @@ func (api *PrivateDebugAPI) TraceChain(ctx context.Context, start, end rpc.Block
 // executes all the transactions contained within. The return value will be one item
 // per transaction, dependent on the requested tracer.
-func (api *PrivateDebugAPI) traceChain(ctx context.Context, start, end *types.Block, config *TraceConfig) (*rpc.Subscription, error) {
+func traceChain(ctx context.Context, eth *Ethereum, start, end *types.Block, config *TraceConfig) (*rpc.Subscription, error) {
```
What do you think about adding some commentary (maybe even package- or file-level) that would explain this extracted method and signature change. Future merges and and conflict-resolving humans :wink: will probably come across this more than once, and I would be nice to be able to get a little refresher on the reasoning and design behind it.

ziogaschr replied:
No need to do this now, as very soon, after the #348 merge, this separation will be removed.  🎉

meowsbits commented on eth/api_tracer_parity.go:
```diff
@@ -56,11 +66,30 @@ func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
 
-	tracer := "callTracerParity"
-	config.Tracer = &tracer
+	if config.Tracer == nil {
```
Rel my other comment at https://github.com/etclabscore/core-geth/pull/345#discussion_r604327234

meowsbits commented on eth/api_tracer_parity.go:
```diff
@@ -56,11 +66,30 @@ func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
 	return config
 }
 
+func decorateParityOutput(res interface{}, config *TraceConfig) interface{} {
```
```suggestion
// decorateParityOutput formats trace results the way Parity does.
// Docs: https://openethereum.github.io/JSONRPC-trace-module
// Example:
/*
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "output": "0x",
    "stateDiff": null,
    "trace": [{
      "action": { ... },
      "result": {
        "gasUsed": "0x0",
        "output": "0x"
      },
      "subtraces": 0,
      "traceAddress": [],
      "type": "call"
    }],
    "vmTrace": null
  }
}
/*
func decorateParityOutput(res interface{}, config *TraceConfig) interface{} {
```

ziogaschr replied:
Resolved by 84bf302

meowsbits commented on eth/tracers/tracer.go:
```diff
@@ -238,6 +238,13 @@ func (dw *dbWrapper) pushObject(vm *duktape.Context) {
+		ctx.PushBoolean(dw.db.Empty(common.BytesToAddress(popSlice(ctx))))
+		return 1
+	})
+	vm.PutPropString(obj, "isEmpty")
```
I think `isEmpty` is a better name, but it looks like other wrapper use the snakeCased method name verbatim. Worth changing to `"empty"` just to adhere to the pattern? At most this is a nitpick...

ziogaschr replied:
No strong preference. `isEmpty` feels to better serve its purpose. Changed to `empty` at 16fb0ec.

meowsbits commented on internal/web3ext/web3ext.go:
```diff
@@ -796,6 +796,18 @@ web3._extend({
+			name: 'callMany',
+			call: 'trace_callMany',
+			params: 3,
+			inputFormatter: [null, null, null]
```
`callMany` uses an array of CallArgs. Should this have its own formatter?  Rel https://github.com/etclabscore/core-geth/pull/345#discussion_r604330305

ziogaschr replied:
Resolved with e0b8ae0

meowsbits committed:
eth: extract condition-to-decorate out of decoration fn

Rel https://github.com/etclabscore/core-geth/pull/345#discussion_r603480090

@ziogaschr, revert this if you hate it ;)

Date: 2021-03-30 16:55:21-05:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
core/vm, eth/tracers: add `CapturePreEVM` to the `Tracer` interface

ziogaschr committed:
core/vm: revert prettyfication commit

It reassigns the `err` var in the if context instead of changing the outer context for the function return

ziogaschr committed:
eth/tracers: rename `isEmpty` to `empty` for db wrapper

ziogaschr committed:
docs: add more on trace-overview docs

ziogaschr committed:
docs: add stateDiff output docs on trace-overview

I removed by mistake the stateDiff parts on previous commit, as I thought I was working on the PR #348.

ziogaschr committed:
internal/web3ext: add better inputFormatters for console’s web3 `trace` calls

ziogaschr committed:
eth: refactor the parity decorate function for clarity

ziogaschr committed:
eth: apply `decorateResponse` on nested txs for callMany

meowsbits commented on eth/api_tracer_parity.go:
```diff
@@ -74,11 +74,34 @@ func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
 
-func decorateParityOutput(res interface{}, config *TraceConfig) interface{} {
+// decorateResponse applies formatting to trace results if needed.
+func decorateResponse(res interface{}, config *TraceConfig) (interface{}, error) {
```
:+1:

meowsbits commented on eth/tracers/internal/tracers/prestate_tracer.js:
```diff
@@ -70,13 +74,12 @@
+
 			this.lookupAccount(log.contract.getAddress(), db);
 		}
+
```
Man this is a lot tidier. Well done!

meowsbits commented on eth/tracers/testdata/state_diff_tracer_big_slow.json:
```diff
@@ -0,0 +1,100 @@
+        "storage": {}
+      }
+    },
+    "config": {
```
A note that using the Mordor config (both configuration format and configuration itself) for test cases limits their applicability to just core-geth; ethereumgo-ethereum won't be able to run these. 

Is it possible to use `&goethereum.ChainConfig` schemas here so they could be more easily shared upstream? And to use an arbitrary chain configuration/genesis that ethereum/go-ethereum would be able to support?

ziogaschr replied:
Great point. Will have a look at this together with the rest stuff discussed yesterday. 
I believe this fits nicely on a new PR and doesn't worth blocking this PR at the moment. 
I agree, that it has to be done, before making an upstream PR.

meowsbits replied:
SGTM!

ziogaschr replied:
Resolved by https://github.com/etclabscore/core-geth/pull/355/commits/208da5589f626f9200ec11f557b373dfc5645e40

meowsbits commented on eth/tracers/tracers_test.go:
```diff
@@ -473,3 +474,161 @@ func BenchmarkCallTracerParity(b *testing.B) {
+		return core.CanTransfer(db, sender, amount)
+	}
+
+	// If the actual transaction would fail, then their is no reason to actually transfer any balance at all
```
```suggestion
	// If the actual transaction would fail, then there is no reason to actually transfer any balance at all
```

meowsbits commented on eth/tracers/tracers_test.go:
```diff
@@ -473,3 +474,161 @@ func BenchmarkCallTracerParity(b *testing.B) {
+	}
+	_, statedb := tests.MakePreState(rawdb.NewMemoryDatabase(), test.Genesis.Alloc, false)
+
+	// Store the truth on wether from acount has enough balance for context usage
```
```suggestion
	// Store the truth on whether from account has enough balance for context usage
```

meowsbits commented on eth/api_tracer.go:
```diff
@@ -802,19 +811,100 @@ func (api *PrivateDebugAPI) TraceTransaction(ctx context.Context, hash common.Ha
+		if config != nil && config.Reexec != nil {
+			reexec = *config.Reexec
+		}
+		_, _, statedb, err = computeTxEnv(eth, block, len(block.Transactions())-1, reexec)
```
IIRC, this `len(block.txs)-1` (vs. `0`) was related to a bug found at ethereum/go-ethereum w/r/t the computed env, or? Just out of curiousity, did report that or they get it fixed?

ziogaschr replied:
Yes, that's indeed a bug that we handled here. From what I recall they handled this fine in the trace package refactor, but I will have to check for sure.

It's on my list of things to refer/PR upstream.

ziogaschr replied:
@meowsbits this has been resolved upstream.

meowsbits commented on eth/tracers/internal/tracers/state_diff_tracer.js:
```diff
@@ -0,0 +1,537 @@
+	includeOpError: function(err) {
+		return err
+			&& err.indexOf('contract address collision') > -1;
+			// && (err.indexOf('insufficient balance for transfer') > -1
```
Dead code or developer clue?

ziogaschr replied:
It's a developer clue on a flow I found on OE code, but I am not sure if reachable on state_diff and I couldn't find a use case or proof that it is needed.

I will remove it as it doesn't look good. Either way, I have some notes in my agenda with regards such things.

ziogaschr replied:
Resolved by f553df4

meowsbits commented on eth/tracers/internal/tracers/state_diff_tracer.js:
```diff
@@ -0,0 +1,537 @@
+		if (this.stateDiff[toAccountHex] !== undefined) {
+			var toAcc = this.stateDiff[toAccountHex];
+
+			// TODO: check if it needs to be added back
```
:tomato: A real pending TODO or DONE?

ziogaschr replied:
Leftover. Has been tested and it can be removed

ziogaschr replied:
Resolved by 523cafe

ziogaschr committed:
eth/tracers: remove commented out code (dead code)

ziogaschr committed:
eth/tracers: remove unneeded code (commented out)

ziogaschr committed:
Fix typos in comments

Co-authored-by: meowsbits <b5c6@protonmail.com>

meowsbits reviewed (approved):
Awesome work.
